module/documents/item.mjs

```diff
@@ -641,6 +644,38 @@ export default class Item5e extends Item {
 
   /* -------------------------------------------- */
 
+  /**
+   * Replace referenced data attributes in the roll formula with values from the provided data.
+   * If the attribute is not found in the provided data, display a warning on the actor.
+   * @param {string} formula           The original formula within which to replace.
+   * @param {object} data              The data object which provides replacements.
+   * @param {object} options
+   * @param {string} options.property  Name of the property to which this formula belongs.
+   * @returns {string}                 Formula with replaced data.
+   */
+  replaceFormulaData(formula, data, { property }) {
```

Comment:
I wonder if, instead of duplicating some of the `Roll.replaceFormulaData` code here, we can instead use the `missing` value to accomplish the same thing. Since it's a string, we could supply a special value to `Roll.replaceFormulaData` that we then look for afterwards:

```js
const replaced = Roll.replaceFormulaData(formula, data, {missing: "$missing$"});
if ( replaced.includes("$missing$") && this.actor ) {
  // Format warning here.
}
return replaced.replaceAll("$missing$", "0");
```
It's a little bit less efficient because we're doing more passes/searches/replaces over the string, but formula strings are pretty short and I think the code ends up a little bit neater as a result.

Reply:
The problem with that method is it would leave no way to access the replaced `@` formula, so I wouldn't be able to include that in the warning.

Reply:
Ah you're right, that would defeat a lot of the usefulness here. This should be on the core system to improve, at least adding a static property for the data formula regex.